src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -70,589 +91,522 @@ public async ValueTask DisposeAsync()
             }
 
             _disposed = true;
-            _queuedCommandsWriter.Complete();
-            await _pipeWriter.CompleteAsync().ConfigureAwait(false);
+
+            if (_pipeWriter != null)
+                await _pipeWriter.CompleteAsync().ConfigureAwait(false);
+            if (_pipeReader != null)
+                await _pipeReader.CompleteAsync().ConfigureAwait(false);
+
+            await _readerLoopTask.ConfigureAwait(false);
         }
         finally
         {
             _semLock.Release();
         }
     }
 
-    public NatsPipeliningWriteProtocolProcessor CreateNatsPipeliningWriteProtocolProcessor(ISocketConnection socketConnection) => new(socketConnection, this, _opts, _counter);
-
-    public ValueTask ConnectAsync(ClientOpts connectOpts, CancellationToken cancellationToken)
+    public async ValueTask ConnectAsync(ClientOpts connectOpts, CancellationToken cancellationToken)
     {
-#pragma warning disable CA2016
-#pragma warning disable VSTHRD103
-        if (!_semLock.Wait(0))
-#pragma warning restore VSTHRD103
-#pragma warning restore CA2016
-        {
-            return ConnectStateMachineAsync(false, connectOpts, cancellationToken);
-        }
-
-        if (_flushTask is { IsCompletedSuccessfully: false })
-        {
-            return ConnectStateMachineAsync(true, connectOpts, cancellationToken);
-        }
-
+        Interlocked.Add(ref _counter.PendingMessages, 1);
```

Comment:
copy/paste error. (Was there an argument about Add() being faster?🤷‍♂️)
Fixed.

Reply:
>  (Was there an argument about Add() being faster?🤷‍♂️)

It _might_ be?

As always devil is in the details, I can't comment for ARM but for x86/x64 I did start to look at the tables on [Agner Fog](https://www.agner.org/optimize/instruction_tables.pdf). However,

 - It has been a minute since I have peeked at these tables
 - They do not list `LOCK INC` 
 - It's entirely possible that the JIT decides to pick the best option regardless...

However digging did lead me to [*this*](https://c9x.me/x86/html/file_module_x86_id_140.html) which claims that INC doesn't update the carry flag compared to `ADD`. Whether that still makes a difference in modern arches, I can't say...

This does almost certainly qualify as a micro opt either way, `.Increment()`/`.Decrement()` feel more readable but am happy with either option.